File: 프로그래머스/lv3/72413. 합승 택시 요금/합승 택시 요금.java
import java.util.*;

class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int answer = 0;
        
        int[][] grp = new int[n+1][n+1];
        
        for(int i = 1; i < n+1; i++){
            Arrays.fill(grp[i], 2000002);
            grp[i][i] = 0;
        }
        for(int[] f: fares){
            grp[f[0]][f[1]] = f[2];
            grp[f[1]][f[0]] = f[2];
        }
        
        for(int k = 1; k < n+1; k++){
            for(int i = 1; i < n+1; i++){
                for(int j = 1; j < n+1; j++){
                    grp[i][j] = Math.min(grp[i][j], grp[i][k] + grp[k][j]);
                }
            }
        }
        
        answer = 3000003;
        for(int i = 0; i < n+1; i++){
            answer = Math.min(answer, grp[s][i] + grp[i][a] + grp[i][b]);
        }
        return answer;
    }
}

File: 프로그래머스/lv3/72413. 합승 택시 요금/합승 택시 요금.py
def solution(n, s, a, b, fares):
    
    grp = []
    for i in range(n+1):
        grp.append([1000010 for _ in range(n+1)])
        
    for f in fares:
        grp[f[0]][f[1]] = f[2]
        grp[f[1]][f[0]] = f[2]
        
    for i in range(1, n+1):
        grp[i][i] = 0
        
    for k in range(1, n+1):
        for st in range(1, n+1):
            for t in range(1, n+1):
                grp[st][t] = min(grp[st][t], grp[st][k] + grp[k][t])
                
    ans = 30000500
    for k in range(1, n+1):
        ans = min(ans, grp[s][k] + grp[k][a] + grp[k][b])
        
    return ans